File: python-sdk/reeeductio/crypto.py
"""
Cryptographic utilities for reeeductio.

Provides Ed25519 signing/verification, hashing, and encoding helpers.
"""

import base64
import hashlib
from dataclasses import dataclass

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF


@dataclass
class Ed25519KeyPair:
    """Ed25519 key pair for signing operations."""

    private_key: bytes  # 32 bytes
    public_key: bytes  # 32 bytes

    def to_user_id(self) -> str:
        """
        Convert public key to user identifier format (44-char URL-safe base64).

        Creates header byte with USER type (0b010100) in first 6 bits and version 0 in last 2 bits.
        The header value 0x50 (0b01010000) encodes to 'U' as the first base64 character.

        Returns:
            44-character URL-safe base64 string starting with 'U'
        """
        # Header: [6 bits: USER type (20 = 0b010100)][2 bits: version (0)]
        header = (0b010100 << 2) | 0  # = 0x50
        typed_key = bytes([header]) + self.public_key
        return base64.urlsafe_b64encode(typed_key).decode("ascii")

    def to_tool_id(self) -> str:
        """
        Convert public key to tool identifier format (44-char URL-safe base64).

        Creates header byte with TOOL type (0b010011) in first 6 bits and version 0 in last 2 bits.
        The header value 0x4C (0b01001100) encodes to 'T' as the first base64 character.

        Returns:
            44-character URL-safe base64 string starting with 'T'
        """
        # Header: [6 bits: TOOL type (19 = 0b010011)][2 bits: version (0)]
        header = (0b010011 << 2) | 0  # = 0x4C
        typed_key = bytes([header]) + self.public_key
        return base64.urlsafe_b64encode(typed_key).decode("ascii")

    def to_space_id(self) -> str:
        """
        Convert public key to space identifier format (44-char URL-safe base64).

        Creates header byte with SPACE type (0b000010) in first 6 bits and version 0 in last 2 bits.
        The header value 0x08 (0b00001000) encodes to 'C' as the first base64 character.

        Returns:
            44-character URL-safe base64 string starting with 'C'
        """
        # Header: [6 bits: SPACE type (2 = 0b000010)][2 bits: version (0)]
        header = (0b000010 << 2) | 0  # = 0x08
        typed_key = bytes([header]) + self.public_key
        return base64.urlsafe_b64encode(typed_key).decode("ascii")

    @classmethod
    def from_typed_public_key(cls, typed_key: str) -> bytes:
        """
        Extract raw public key bytes from typed identifier.

        Supports identifiers for USER, TOOL, or SPACE types. The header byte is
        validated but not checked for specific type - any public key type is accepted.

        Args:
            typed_key: 44-character URL-safe base64 string with typed header

        Returns:
            32-byte raw Ed25519 public key

        Raises:
            ValueError: If identifier format is invalid or not a public key type
        """
        if len(typed_key) != 44:
            raise ValueError(f"Typed identifier must be 44 characters, got {len(typed_key)}")

        # Decode (44 base64 chars = 33 bytes: 1 header + 32 data)
        decoded = base64.urlsafe_b64decode(typed_key)

        if len(decoded) != 33:
            raise ValueError(f"Decoded identifier must be 33 bytes, got {len(decoded)}")

        # Extract type from header byte
        header = decoded[0]
        type_bits = (header >> 2) & 0b111111

        # Validate it's a public key type (USER=20, TOOL=19, SPACE=2)
        valid_types = {0b010100, 0b010011, 0b000010}  # USER, TOOL, SPACE
        if type_bits not in valid_types:
            raise ValueError(f"Identifier is not a USER, TOOL, or SPACE type: {type_bits:06b}")

        # Return the 32-byte public key (skip header)
        return decoded[1:]


def get_identifier_type(typed_key: str) -> str:
    """
    Get the type of a typed identifier.

    Args:
        typed_key: 44-character URL-safe base64 string with typed header

    Returns:
        Type name: "USER", "TOOL", "SPACE", "MESSAGE", or "BLOB"

    Raises:
        ValueError: If identifier format is invalid
    """
    if len(typed_key) != 44:
        raise ValueError(f"Typed identifier must be 44 characters, got {len(typed_key)}")

    decoded = base64.urlsafe_b64decode(typed_key)
    if len(decoded) != 33:
        raise ValueError(f"Decoded identifier must be 33 bytes, got {len(decoded)}")

    header = decoded[0]
    type_bits = (header >> 2) & 0b111111

    type_map = {
        0b010100: "USER",    # 20
        0b010011: "TOOL",    # 19
        0b000010: "SPACE",   # 2
        0b001100: "MESSAGE", # 12
        0b000001: "BLOB",    # 1
    }

    type_name = type_map.get(type_bits)
    if type_name is None:
        raise ValueError(f"Unknown identifier type: {type_bits:06b}")

    return type_name


def generate_keypair() -> Ed25519KeyPair:
    """
    Generate a new Ed25519 key pair.

    Returns:
        Ed25519KeyPair with private and public keys
    """
    private_key = Ed25519PrivateKey.generate()
    public_key = private_key.public_key()

    # Extract raw bytes (32 bytes each)
    private_bytes = private_key.private_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PrivateFormat.Raw,
        encryption_algorithm=serialization.NoEncryption(),
    )
    public_bytes = public_key.public_bytes(encoding=serialization.Encoding.Raw, format=serialization.PublicFormat.Raw)

    return Ed25519KeyPair(private_key=private_bytes, public_key=public_bytes)


def sign_data(data: bytes, private_key: bytes) -> bytes:
    """
    Sign data using Ed25519.

    Args:
        data: Data to sign
        private_key: 32-byte Ed25519 private key

    Returns:
        64-byte signature
    """
    # Load private key from raw bytes
    private_key_obj = Ed25519PrivateKey.from_private_bytes(private_key)
    return private_key_obj.sign(data)


def verify_signature(data: bytes, signature: bytes, public_key: bytes) -> bool:
    """
    Verify Ed25519 signature.

    Args:
        data: Data that was signed
        signature: 64-byte signature
        public_key: 32-byte Ed25519 public key

    Returns:
        True if signature is valid, False otherwise
    """
    try:
        public_key_obj = Ed25519PublicKey.from_public_bytes(public_key)
        public_key_obj.verify(signature, data)
        return True
    except InvalidSignature:
        return False
    except Exception:
        return False


def compute_hash(data: bytes) -> bytes:
    """
    Compute SHA256 hash of data.

    Args:
        data: Data to hash

    Returns:
        32-byte SHA256 hash
    """
    return hashlib.sha256(data).digest()


def to_message_id(hash_bytes: bytes) -> str:
    """
    Convert hash to message identifier format (44-char URL-safe base64).

    Creates header byte with MESSAGE type (0b001100) in first 6 bits and version 0 in last 2 bits.
    The header value 0x30 (0b00110000) encodes to 'M' as the first base64 character.

    Args:
        hash_bytes: 32-byte SHA256 hash

    Returns:
        44-character URL-safe base64 string starting with 'M'
    """
    if len(hash_bytes) != 32:
        raise ValueError(f"Hash must be exactly 32 bytes, got {len(hash_bytes)}")

    # Header: [6 bits: MESSAGE type (12 = 0b001100)][2 bits: version (0)]
    header = (0b001100 << 2) | 0  # = 0x30
    typed_id = bytes([header]) + hash_bytes
    return base64.urlsafe_b64encode(typed_id).decode("ascii")


def to_blob_id(hash_bytes: bytes) -> str:
    """
    Convert hash to blob identifier format (44-char URL-safe base64).

    Creates header byte with BLOB type (0b000001) in first 6 bits and version 0 in last 2 bits.
    The header value 0x04 (0b00000100) encodes to 'B' as the first base64 character.

    Args:
        hash_bytes: 32-byte SHA256 hash

    Returns:
        44-character URL-safe base64 string starting with 'B'
    """
    if len(hash_bytes) != 32:
        raise ValueError(f"Hash must be exactly 32 bytes, got {len(hash_bytes)}")

    # Header: [6 bits: BLOB type (1 = 0b000001)][2 bits: version (0)]
    header = (0b000001 << 2) | 0  # = 0x04
    typed_id = bytes([header]) + hash_bytes
    return base64.urlsafe_b64encode(typed_id).decode("ascii")


def encode_base64(data: bytes | str) -> str:
    """
    Encode data as standard base64 string.

    Args:
        data: Bytes or string to encode

    Returns:
        Base64-encoded string
    """
    if isinstance(data, str):
        data = data.encode("utf-8")
    return base64.b64encode(data).decode("ascii")


def decode_base64(data: str) -> bytes:
    """
    Decode base64 string to bytes.

    Args:
        data: Base64-encoded string

    Returns:
        Decoded bytes
    """
    return base64.b64decode(data)


def encode_urlsafe_base64(data: bytes | str) -> str:
    """
    Encode data as URL-safe base64 string (no padding).

    Args:
        data: Bytes or string to encode

    Returns:
        URL-safe base64-encoded string without padding
    """
    if isinstance(data, str):
        data = data.encode("utf-8")
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def decode_urlsafe_base64(data: str) -> bytes:
    """
    Decode URL-safe base64 string to bytes.

    Args:
        data: URL-safe base64-encoded string

    Returns:
        Decoded bytes
    """
    # Add padding if needed
    padding = 4 - (len(data) % 4)
    if padding != 4:
        data += "=" * padding
    return base64.urlsafe_b64decode(data)


def derive_key(root_key: bytes, info: str, length: int = 32) -> bytes:
    """
    Derive a key using HKDF-SHA256.

    Args:
        root_key: Root key material (typically 32 bytes)
        info: Context/purpose string for key derivation
        length: Desired output key length in bytes (default: 32)

    Returns:
        Derived key of specified length

    Example:
        >>> root = os.urandom(32)
        >>> message_key = derive_key(root, "message key")
        >>> blob_key = derive_key(root, "blob key")
    """
    hkdf = HKDF(
        algorithm=hashes.SHA256(),
        length=length,
        salt=None,
        info=info.encode("utf-8"),
    )
    return hkdf.derive(root_key)


def encrypt_aes_gcm(plaintext: bytes, key: bytes, associated_data: bytes | None = None) -> bytes:
    """
    Encrypt data using AES-GCM-256.

    The output format is: IV (12 bytes) + ciphertext + tag (16 bytes)

    Args:
        plaintext: Data to encrypt
        key: 32-byte AES-256 key
        associated_data: Optional additional authenticated data (AAD)

    Returns:
        Encrypted data with IV + ciphertext + tag concatenated

    Raises:
        ValueError: If key is not 32 bytes
    """
    if len(key) != 32:
        raise ValueError(f"AES-256 key must be exactly 32 bytes, got {len(key)}")

    # Generate random 12-byte IV (recommended for GCM)
    import os
    iv = os.urandom(12)

    # Create cipher and encrypt
    aesgcm = AESGCM(key)
    ciphertext_with_tag = aesgcm.encrypt(iv, plaintext, associated_data)

    # Return IV + ciphertext + tag
    return iv + ciphertext_with_tag


def decrypt_aes_gcm(encrypted: bytes, key: bytes, associated_data: bytes | None = None) -> bytes:
    """
    Decrypt AES-GCM-256 encrypted data.

    Expects input format: IV (12 bytes) + ciphertext + tag (16 bytes)

    Args:
        encrypted: Encrypted data (IV + ciphertext + tag)
        key: 32-byte AES-256 key
        associated_data: Optional additional authenticated data (AAD), must match encryption

    Returns:
        Decrypted plaintext

    Raises:
        ValueError: If key is not 32 bytes or encrypted data is too short
        cryptography.exceptions.InvalidTag: If authentication fails (wrong key, corrupted data, or wrong AAD)
    """
    if len(key) != 32:
        raise ValueError(f"AES-256 key must be exactly 32 bytes, got {len(key)}")

    # Minimum size is 12 (IV) + 16 (tag) = 28 bytes
    if len(encrypted) < 28:
        raise ValueError(f"Encrypted data too short, must be at least 28 bytes, got {len(encrypted)}")

    # Extract IV (first 12 bytes)
    iv = encrypted[:12]

    # Rest is ciphertext + tag (handled by AESGCM.decrypt)
    ciphertext_with_tag = encrypted[12:]

    # Decrypt and verify
    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(iv, ciphertext_with_tag, associated_data)

    return plaintext

File: typescript-sdk/src/crypto.ts
/**
 * Cryptographic utilities for reeeductio.
 *
 * Provides Ed25519 signing/verification, AES-GCM encryption, hashing, and encoding helpers.
 */

import * as ed from '@noble/ed25519';
import { sha256 } from '@noble/hashes/sha256';
import { hkdf } from '@noble/hashes/hkdf';
import { gcm } from '@noble/ciphers/aes.js';
import { randomBytes } from '@noble/ciphers/utils.js';
import { IdType, type KeyPair } from './types.js';

/**
 * Generate a new Ed25519 key pair.
 */
export async function generateKeyPair(): Promise<KeyPair> {
  const privateKey = randomBytes(32);
  const publicKey = await ed.getPublicKeyAsync(privateKey);
  return { privateKey, publicKey };
}

/**
 * Sign data using Ed25519.
 *
 * @param data - Data to sign
 * @param privateKey - 32-byte Ed25519 private key
 * @returns 64-byte signature
 */
export async function signData(data: Uint8Array, privateKey: Uint8Array): Promise<Uint8Array> {
  return ed.signAsync(data, privateKey);
}

/**
 * Verify Ed25519 signature.
 *
 * @param data - Data that was signed
 * @param signature - 64-byte signature
 * @param publicKey - 32-byte Ed25519 public key
 * @returns True if signature is valid
 */
export async function verifySignature(
  data: Uint8Array,
  signature: Uint8Array,
  publicKey: Uint8Array
): Promise<boolean> {
  try {
    return await ed.verifyAsync(signature, data, publicKey);
  } catch {
    return false;
  }
}

/**
 * Compute SHA256 hash of data.
 *
 * @param data - Data to hash
 * @returns 32-byte SHA256 hash
 */
export function computeHash(data: Uint8Array): Uint8Array {
  return sha256(data);
}

/**
 * Derive a key using HKDF-SHA256.
 *
 * @param rootKey - Root key material (typically 32 bytes)
 * @param info - Context/purpose string for key derivation
 * @param length - Desired output key length in bytes (default: 32)
 * @returns Derived key of specified length
 */
export function deriveKey(rootKey: Uint8Array, info: string, length: number = 32): Uint8Array {
  const infoBytes = new TextEncoder().encode(info);
  return hkdf(sha256, rootKey, undefined, infoBytes, length);
}

/**
 * Encrypt data using AES-GCM-256.
 *
 * The output format is: IV (12 bytes) + ciphertext + tag (16 bytes)
 *
 * @param plaintext - Data to encrypt
 * @param key - 32-byte AES-256 key
 * @param associatedData - Optional additional authenticated data (AAD)
 * @returns Encrypted data with IV + ciphertext + tag concatenated
 */
export function encryptAesGcm(
  plaintext: Uint8Array,
  key: Uint8Array,
  associatedData?: Uint8Array
): Uint8Array {
  if (key.length !== 32) {
    throw new Error(`AES-256 key must be exactly 32 bytes, got ${key.length}`);
  }

  // Generate random 12-byte IV (recommended for GCM)
  const iv = randomBytes(12);

  // Create cipher and encrypt
  const cipher = gcm(key, iv, associatedData);
  const ciphertextWithTag = cipher.encrypt(plaintext);

  // Return IV + ciphertext + tag
  const result = new Uint8Array(iv.length + ciphertextWithTag.length);
  result.set(iv, 0);
  result.set(ciphertextWithTag, iv.length);
  return result;
}

/**
 * Decrypt AES-GCM-256 encrypted data.
 *
 * Expects input format: IV (12 bytes) + ciphertext + tag (16 bytes)
 *
 * @param encrypted - Encrypted data (IV + ciphertext + tag)
 * @param key - 32-byte AES-256 key
 * @param associatedData - Optional additional authenticated data (AAD)
 * @returns Decrypted plaintext
 */
export function decryptAesGcm(
  encrypted: Uint8Array,
  key: Uint8Array,
  associatedData?: Uint8Array
): Uint8Array {
  if (key.length !== 32) {
    throw new Error(`AES-256 key must be exactly 32 bytes, got ${key.length}`);
  }

  // Minimum size is 12 (IV) + 16 (tag) = 28 bytes
  if (encrypted.length < 28) {
    throw new Error(`Encrypted data too short, must be at least 28 bytes, got ${encrypted.length}`);
  }

  // Extract IV (first 12 bytes)
  const iv = encrypted.slice(0, 12);

  // Rest is ciphertext + tag
  const ciphertextWithTag = encrypted.slice(12);

  // Decrypt and verify
  const cipher = gcm(key, iv, associatedData);
  return cipher.decrypt(ciphertextWithTag);
}

/**
 * Encode bytes as standard base64 string.
 */
export function encodeBase64(data: Uint8Array): string {
  if (typeof Buffer !== 'undefined') {
    return Buffer.from(data).toString('base64');
  }
  // Browser environment
  return btoa(String.fromCharCode(...data));
}

/**
 * Decode base64 string to bytes.
 */
export function decodeBase64(data: string): Uint8Array {
  if (typeof Buffer !== 'undefined') {
    return new Uint8Array(Buffer.from(data, 'base64'));
  }
  // Browser environment
  return new Uint8Array(atob(data).split('').map((c) => c.charCodeAt(0)));
}

/**
 * Encode bytes as URL-safe base64 string (with padding).
 */
export function encodeUrlSafeBase64(data: Uint8Array): string {
  const base64 = encodeBase64(data);
  return base64.replace(/\+/g, '-').replace(/\//g, '_');
}

/**
 * Decode URL-safe base64 string to bytes.
 */
export function decodeUrlSafeBase64(data: string): Uint8Array {
  // Convert URL-safe to standard base64
  let base64 = data.replace(/-/g, '+').replace(/_/g, '/');
  // Add padding if needed
  const padding = 4 - (base64.length % 4);
  if (padding !== 4) {
    base64 += '='.repeat(padding);
  }
  return decodeBase64(base64);
}

/**
 * Convert public key to user identifier format (44-char URL-safe base64).
 *
 * Creates header byte with USER type (0b010100) in first 6 bits and version 0 in last 2 bits.
 * The header value 0x50 (0b01010000) encodes to 'U' as the first base64 character.
 *
 * @param publicKey - 32-byte Ed25519 public key
 * @returns 44-character URL-safe base64 string starting with 'U'
 */
export function toUserId(publicKey: Uint8Array): string {
  if (publicKey.length !== 32) {
    throw new Error(`Public key must be exactly 32 bytes, got ${publicKey.length}`);
  }
  // Header: [6 bits: USER type (20 = 0b010100)][2 bits: version (0)]
  const header = (IdType.USER << 2) | 0; // = 0x50
  const typed = new Uint8Array(33);
  typed[0] = header;
  typed.set(publicKey, 1);
  return encodeUrlSafeBase64(typed);
}

/**
 * Convert public key to tool identifier format (44-char URL-safe base64).
 *
 * Creates header byte with TOOL type (0b010011) in first 6 bits and version 0 in last 2 bits.
 * The header value 0x4C (0b01001100) encodes to 'T' as the first base64 character.
 *
 * @param publicKey - 32-byte Ed25519 public key
 * @returns 44-character URL-safe base64 string starting with 'T'
 */
export function toToolId(publicKey: Uint8Array): string {
  if (publicKey.length !== 32) {
    throw new Error(`Public key must be exactly 32 bytes, got ${publicKey.length}`);
  }
  // Header: [6 bits: TOOL type (19 = 0b010011)][2 bits: version (0)]
  const header = (IdType.TOOL << 2) | 0; // = 0x4C
  const typed = new Uint8Array(33);
  typed[0] = header;
  typed.set(publicKey, 1);
  return encodeUrlSafeBase64(typed);
}

/**
 * Convert public key to space identifier format (44-char URL-safe base64).
 *
 * Creates header byte with SPACE type (0b000010) in first 6 bits and version 0 in last 2 bits.
 * The header value 0x08 (0b00001000) encodes to 'C' as the first base64 character.
 *
 * @param publicKey - 32-byte Ed25519 public key
 * @returns 44-character URL-safe base64 string starting with 'C'
 */
export function toSpaceId(publicKey: Uint8Array): string {
  if (publicKey.length !== 32) {
    throw new Error(`Public key must be exactly 32 bytes, got ${publicKey.length}`);
  }
  // Header: [6 bits: SPACE type (2 = 0b000010)][2 bits: version (0)]
  const header = (IdType.SPACE << 2) | 0; // = 0x08
  const typed = new Uint8Array(33);
  typed[0] = header;
  typed.set(publicKey, 1);
  return encodeUrlSafeBase64(typed);
}

/**
 * Convert hash to message identifier format (44-char URL-safe base64).
 *
 * Creates header byte with MESSAGE type (0b001100) in first 6 bits and version 0 in last 2 bits.
 * The header value 0x30 (0b00110000) encodes to 'M' as the first base64 character.
 *
 * @param hashBytes - 32-byte SHA256 hash
 * @returns 44-character URL-safe base64 string starting with 'M'
 */
export function toMessageId(hashBytes: Uint8Array): string {
  if (hashBytes.length !== 32) {
    throw new Error(`Hash must be exactly 32 bytes, got ${hashBytes.length}`);
  }
  // Header: [6 bits: MESSAGE type (12 = 0b001100)][2 bits: version (0)]
  const header = (IdType.MESSAGE << 2) | 0; // = 0x30
  const typed = new Uint8Array(33);
  typed[0] = header;
  typed.set(hashBytes, 1);
  return encodeUrlSafeBase64(typed);
}

/**
 * Convert hash to blob identifier format (44-char URL-safe base64).
 *
 * Creates header byte with BLOB type (0b000001) in first 6 bits and version 0 in last 2 bits.
 * The header value 0x04 (0b00000100) encodes to 'B' as the first base64 character.
 *
 * @param hashBytes - 32-byte SHA256 hash
 * @returns 44-character URL-safe base64 string starting with 'B'
 */
export function toBlobId(hashBytes: Uint8Array): string {
  if (hashBytes.length !== 32) {
    throw new Error(`Hash must be exactly 32 bytes, got ${hashBytes.length}`);
  }
  // Header: [6 bits: BLOB type (1 = 0b000001)][2 bits: version (0)]
  const header = (IdType.BLOB << 2) | 0; // = 0x04
  const typed = new Uint8Array(33);
  typed[0] = header;
  typed.set(hashBytes, 1);
  return encodeUrlSafeBase64(typed);
}

/**
 * Get the type of a typed identifier.
 *
 * @param typedId - 44-character URL-safe base64 string with typed header
 * @returns The identifier type
 */
export function getIdentifierType(typedId: string): IdType {
  if (typedId.length !== 44) {
    throw new Error(`Typed identifier must be 44 characters, got ${typedId.length}`);
  }

  const decoded = decodeUrlSafeBase64(typedId);
  if (decoded.length !== 33) {
    throw new Error(`Decoded identifier must be 33 bytes, got ${decoded.length}`);
  }

  const header = decoded[0];
  const typeBits = (header >> 2) & 0b111111;

  const validTypes = [IdType.USER, IdType.TOOL, IdType.SPACE, IdType.MESSAGE, IdType.BLOB];
  if (!validTypes.includes(typeBits)) {
    throw new Error(`Unknown identifier type: ${typeBits.toString(2).padStart(6, '0')}`);
  }

  return typeBits as IdType;
}

/**
 * Extract raw bytes (public key or hash) from typed identifier.
 *
 * @param typedId - 44-character URL-safe base64 string with typed header
 * @returns 32-byte raw data (public key or hash)
 */
export function extractFromTypedId(typedId: string): Uint8Array {
  if (typedId.length !== 44) {
    throw new Error(`Typed identifier must be 44 characters, got ${typedId.length}`);
  }

  const decoded = decodeUrlSafeBase64(typedId);
  if (decoded.length !== 33) {
    throw new Error(`Decoded identifier must be 33 bytes, got ${decoded.length}`);
  }

  // Return the 32-byte data (skip header)
  return decoded.slice(1);
}

/**
 * Extract public key from a USER, TOOL, or SPACE identifier.
 *
 * @param typedId - 44-character URL-safe base64 identifier
 * @returns 32-byte Ed25519 public key
 */
export function extractPublicKey(typedId: string): Uint8Array {
  const idType = getIdentifierType(typedId);
  const validPublicKeyTypes = [IdType.USER, IdType.TOOL, IdType.SPACE];

  if (!validPublicKeyTypes.includes(idType)) {
    throw new Error(`Identifier is not a public key type (USER, TOOL, or SPACE)`);
  }

  return extractFromTypedId(typedId);
}

/**
 * Concatenate multiple Uint8Arrays.
 */
export function concatBytes(...arrays: Uint8Array[]): Uint8Array {
  const totalLength = arrays.reduce((sum, arr) => sum + arr.length, 0);
  const result = new Uint8Array(totalLength);
  let offset = 0;
  for (const arr of arrays) {
    result.set(arr, offset);
    offset += arr.length;
  }
  return result;
}

/**
 * Convert a string to UTF-8 bytes.
 */
export function stringToBytes(str: string): Uint8Array {
  return new TextEncoder().encode(str);
}

/**
 * Convert UTF-8 bytes to a string.
 */
export function bytesToString(bytes: Uint8Array): string {
  return new TextDecoder().decode(bytes);
}
